### ISSUE TITLE
Validate Tekton works with GKE Autopilot

### ISSUE THREAD

**Issue by imjasonh** [tektoncd/pipeline#3798]
GKE Autopilot is a new mode for GKE which locks down certain aspects of the cluster, in exchange for a more managed environment, and billing based on pod resource requests instead of node reservations. Someone should make sure Tekton works well with it, or at least identify where it doesn't and document those.

Looking through the overview (https://cloud.google.com/kubernetes-engine/docs/concepts/autopilot-overview) there aren't many things locked down that we might rely on, but a few might be problematic:

[Webhook limitations](https://cloud.google.com/kubernetes-engine/docs/concepts/autopilot-overview#webhooks_limitations):
> You cannot create custom mutating admission webhooks for Autopilot clusters, but you can create custom validating webhooks.

Tekton uses mutating webhooks to set defaults. There's also no mention of conversion webhooks, which Tekton uses.

[Pod affinity and anti-affinity](https://cloud.google.com/kubernetes-engine/docs/concepts/autopilot-overview#pod_affinity_and_anti-affinity):
> Pod affinity is limited for use only with the following keys: `topology.kubernetes.io/region`, `topology.kubernetes.io/zone`, `failure-domain.beta.kubernetes.io/region`, and `failure-domain.beta.kubernetes.io/zone`.

This might affect the Affinity Assistant

[Allowable resource ranges](https://cloud.google.com/kubernetes-engine/docs/concepts/autopilot-overview#allowable_resource_ranges):
> The minimum value is 250 milliCPU (mCPU).

> Containers with no resource requests will default to the standard minimums of 500 mCPU and 1 GiB memory.

We should set reasonable resource requests for controller and webhook deployments, especially if we think we should request lower than the default.

We might find out other things don't work as expected as well. If things already work fine with Autopilot, we should document that somewhere too.

/kind documentation

**Comment by nikhil-thomas** [comment]
🧑‍💻 🎉 

in short, 
installing pipeline `v0.24.1` on GKE Autopilot cluster:
```
Error from server (Forbidden): error when retrieving current configuration of:
Resource: "admissionregistration.k8s.io/v1, Resource=mutatingwebhookconfigurations", GroupVersionKind: "admissionregistration.k8s.io/v1, Kind=MutatingWebhookConfiguration"
Name: "webhook.pipeline.tekton.dev", Namespace: ""
from server for: "https://github.com/tektoncd/pipeline/releases/download/v0.24.1/release.notags.yaml": mutatingwebhookconfigurations.admissionregistration.k8s.io "webhook.pi
peline.tekton.dev" is forbidden: User "nikhilthomas1@gmail.com" cannot get resource "mutatingwebhookconfigurations" in API group "admissionregistration.k8s.io" at the cluste
r scope: GKEAutopilot authz: cluster scoped resource "mutatingwebhookconfigurations/" is managed and access is denied
```

@imjasonh @vdemeester @bobcatfish 

installation log
```
kubectl apply -f https://github.com/tektoncd/pipeline/releases/download/v0.24.1/release.notags.yaml                                                                          

namespace/tekton-pipelines created                                                                                                                                           
podsecuritypolicy.policy/tekton-pipelines created                                                                                                                            
clusterrole.rbac.authorization.k8s.io/tekton-pipelines-controller-cluster-access created                                                                                     
clusterrole.rbac.authorization.k8s.io/tekton-pipelines-controller-tenant-access created                                                                                      
clusterrole.rbac.authorization.k8s.io/tekton-pipelines-webhook-cluster-access created                                                                                        
role.rbac.authorization.k8s.io/tekton-pipelines-controller created                                                                                                           
role.rbac.authorization.k8s.io/tekton-pipelines-webhook created                                                                                                              
role.rbac.authorization.k8s.io/tekton-pipelines-leader-election created                                                                                                      
serviceaccount/tekton-pipelines-controller created                                                                                                                           
serviceaccount/tekton-pipelines-webhook created                                                                                                                              
clusterrolebinding.rbac.authorization.k8s.io/tekton-pipelines-controller-cluster-access created
clusterrolebinding.rbac.authorization.k8s.io/tekton-pipelines-controller-tenant-access created
clusterrolebinding.rbac.authorization.k8s.io/tekton-pipelines-webhook-cluster-access created
rolebinding.rbac.authorization.k8s.io/tekton-pipelines-controller created
rolebinding.rbac.authorization.k8s.io/tekton-pipelines-webhook created
rolebinding.rbac.authorization.k8s.io/tekton-pipelines-controller-leaderelection created
rolebinding.rbac.authorization.k8s.io/tekton-pipelines-webhook-leaderelection created
customresourcedefinition.apiextensions.k8s.io/clustertasks.tekton.dev created
customresourcedefinition.apiextensions.k8s.io/conditions.tekton.dev created
customresourcedefinition.apiextensions.k8s.io/pipelines.tekton.dev created
customresourcedefinition.apiextensions.k8s.io/pipelineruns.tekton.dev created
customresourcedefinition.apiextensions.k8s.io/pipelineresources.tekton.dev created
customresourcedefinition.apiextensions.k8s.io/runs.tekton.dev created
customresourcedefinition.apiextensions.k8s.io/tasks.tekton.dev created
customresourcedefinition.apiextensions.k8s.io/taskruns.tekton.dev created
secret/webhook-certs created
validatingwebhookconfiguration.admissionregistration.k8s.io/validation.webhook.pipeline.tekton.dev created
validatingwebhookconfiguration.admissionregistration.k8s.io/config.webhook.pipeline.tekton.dev created
clusterrole.rbac.authorization.k8s.io/tekton-aggregate-edit created
clusterrole.rbac.authorization.k8s.io/tekton-aggregate-view created
configmap/config-artifact-bucket created
configmap/config-artifact-pvc created
configmap/config-defaults created
configmap/feature-flags created
configmap/config-leader-election created
configmap/config-logging created
configmap/config-observability created
configmap/config-registry-cert created
deployment.apps/tekton-pipelines-controller created
service/tekton-pipelines-controller created                                                                                                                                  
horizontalpodautoscaler.autoscaling/tekton-pipelines-webhook created                                                                                                         
deployment.apps/tekton-pipelines-webhook created                   
service/tekton-pipelines-webhook created                                              
Error from server (Forbidden): error when retrieving current configuration of:
Resource: "admissionregistration.k8s.io/v1, Resource=mutatingwebhookconfigurations", GroupVersionKind: "admissionregistration.k8s.io/v1, Kind=MutatingWebhookConfiguration"
Name: "webhook.pipeline.tekton.dev", Namespace: ""
from server for: "https://github.com/tektoncd/pipeline/releases/download/v0.24.1/release.notags.yaml": mutatingwebhookconfigurations.admissionregistration.k8s.io "webhook.pi
peline.tekton.dev" is forbidden: User "nikhilthomas1@gmail.com" cannot get resource "mutatingwebhookconfigurations" in API group "admissionregistration.k8s.io" at the cluste
r scope: GKEAutopilot authz: cluster scoped resource "mutatingwebhookconfigurations/" is managed and access is denied
```

**Comment by imjasonh** [comment]
Thanks for trying it out @nikhil-thomas 

Do e2e tests pass against the autopilot cluster?

In theory we could redo any webhook-initiated mutations on a resource's first reconcilation (which we should maybe do anyway, and maybe already do), which would make the mutating webhook a nice-to-have optimization.

**Comment by nikhil-thomas** [comment]
i haven't tried. I shall add it to my list and post an update soon. 🧑‍💻

**Comment by aemengo** [comment]
I am also running into this issue 👋🏾

**Comment by tlawrie** [comment]
I am also running into this issue. Happy to help out testing if I can.

**Comment by imjasonh** [comment]
If the lack of webhooks is indeed the only issue (big if), I think this boils down to two issues:

1. Allow installation to succeed even if setting up the webhooks fails -- this is probably an Operator issue.
2. Allow Tekton to work correctly if the webhooks aren't available -- detect this scenario and set defaults etc on the first reconciliation, then things should proceed as normal.

And a secret third:

3. Cover this scenario in automated tests, so we know we don't break this in future changes.

**Comment by tlawrie** [comment]
Seeing as my background is Java and my Go is basic / still learning. Can I help with 1 or 2? I was using the helm chart, not sure if that's using the operator under the covers or whether its slightly different.

**Comment by jsravn** [comment]
I thought I could get away with this autopilot restriction, but it seems tekton-pipelines-webhook doesn't start up correctly due to the MutatingWebhookConfiguration. Other things seem to work okay though.

**Comment by vdemeester** [comment]
> I thought I could get away with this autopilot restriction, but it seems tekton-pipelines-webhook doesn't start up correctly due to the MutatingWebhookConfiguration. Other things seem to work okay though.

The trick is that we use the mutating webhook for our default values set or our auto conversion from v1alpha1 to v1beta1, ..

**Comment by imjasonh** [comment]
As a test, I created a kind cluster, installed latest Tekton, disabled webhooks:

```
kubectl delete validatingwebhookconfigurations.admissionregistration.k8s.io validation.webhook.pipeline.tekton.dev
kubectl delete validatingwebhookconfigurations.admissionregistration.k8s.io config.webhook.pipeline.tekton.dev
kubectl delete mutatingwebhookconfigurations.admissionregistration.k8s.io webhook.pipeline.tekton.dev
kubectl delete svc -n tekton-pipelines tekton-pipelines-webhook
```

...and ran the e2e tests with:

```
SYSTEM_NAMESPACE=tekton-pipelines go test -tags=e2e ./test > e2e-testlog.txt
```

logs attached: [e2e-testlog.txt](https://github.com/tektoncd/pipeline/files/7280018/e2e-testlog.txt)

Only a few tests failed, mainly because they expected to have the webhook block an invalid request. If we loosened these tests to handle async validation on the first reconcile loop, they should pass.

**Comment by imjasonh** [comment]
https://cloud.google.com/kubernetes-engine/docs/concepts/autopilot-overview

> In GKE version 1.21 and later, you can also create mutating dynamic admission webhooks. However, Autopilot modifies mutating webhooks objects to add a namespace selector which excludes the resources in managed namespaces (e.g. kube-system) from being intercepted.

tl;dr: it should be much easier to test on GKE Autopilot >1.21 now.

**Comment by joeyslalom** [comment]
Tried tekton (v0.33.0) with Autopilot (1.21.6-gke.1500) with a pipeline with two tasks: `git-clone` and `kaniko` (both v0.5).  My roadbumps, well predicted by @imjasonh 
* Needed to update the firewall rule to allow port 8443 for the admission webhook
* Had to disable affinity assistant, else `admission webhook "policycontrollerv2.common-webhooks.networking.gke.io" denied the request`
* Had to increase ephermal-storage via a `LimitRange`

Generally I should also say with Autopilot scheduling and using `kaniko` to build a 1Gi image this was slow.  PipelineRuns were all 10min+

**Comment by imjasonh** [comment]
I've played with Tekton on GKE Autopilot, including with GKE Spot Pods, and everything in my tests seems to work fine. There might be room for some official GCP-authored-and-hosted doc about how to make them work best together, but that's probably out of scope for Tekton.

If someone finds a bug that makes it not work with GKE Autopilot, please let us know. Until then:

/close
